auth0/src/main/java/com/auth0/android/authentication/storage/CredentialsManager.java

```diff
@@ -100,34 +93,65 @@ public void saveCredentials(@NonNull Credentials credentials) {
      *
      * @param callback the callback that will receive a valid {@link Credentials} or the {@link CredentialsManagerException}.
      */
-    public void getCredentials(@NonNull final BaseCallback<Credentials, CredentialsManagerException> callback) {
+    public void getCredentials(@NonNull BaseCallback<Credentials, CredentialsManagerException> callback) {
+        getCredentials(null, 0, callback);
+    }
+
+    /**
+     * Retrieves the credentials from the storage and refresh them if they have already expired.
+     * It will fail with {@link CredentialsManagerException} if the saved access_token or id_token is null,
+     * or if the tokens have already expired and the refresh_token is null.
+     *
+     * @param scope    the scope to request for the access token. If null is passed, the previous scope will be kept.
+     * @param minTtl   the minimum time in seconds that both the access token and id token should last before expiration.
+     * @param callback the callback that will receive a valid {@link Credentials} or the {@link CredentialsManagerException}.
+     */
+    public void getCredentials(@Nullable String scope, final int minTtl, @NonNull final BaseCallback<Credentials, CredentialsManagerException> callback) {
         String accessToken = storage.retrieveString(KEY_ACCESS_TOKEN);
         final String refreshToken = storage.retrieveString(KEY_REFRESH_TOKEN);
         String idToken = storage.retrieveString(KEY_ID_TOKEN);
         String tokenType = storage.retrieveString(KEY_TOKEN_TYPE);
         Long expiresAt = storage.retrieveLong(KEY_EXPIRES_AT);
-        String scope = storage.retrieveString(KEY_SCOPE);
+        String storedScope = storage.retrieveString(KEY_SCOPE);
         Long cacheExpiresAt = storage.retrieveLong(KEY_CACHE_EXPIRES_AT);
         if (cacheExpiresAt == null) {
             cacheExpiresAt = expiresAt;
         }
 
-        if (isEmpty(accessToken) && isEmpty(idToken) || expiresAt == null) {
+        boolean hasEmptyCredentials = isEmpty(accessToken) && isEmpty(idToken) || expiresAt == null;
+        if (hasEmptyCredentials) {
             callback.onFailure(new CredentialsManagerException("No Credentials were previously set."));
             return;
         }
-        if (cacheExpiresAt > getCurrentTimeInMillis()) {
-            callback.onSuccess(recreateCredentials(idToken, accessToken, tokenType, refreshToken, new Date(expiresAt), scope));
+
+        boolean willExpire = willExpire(cacheExpiresAt, minTtl);
+        boolean scopeChanged = hasScopeChanged(storedScope, scope);
+
+        if (!willExpire && !scopeChanged) {
+            callback.onSuccess(recreateCredentials(idToken, accessToken, tokenType, refreshToken, new Date(expiresAt), storedScope));
             return;
         }
         if (refreshToken == null) {
             callback.onFailure(new CredentialsManagerException("Credentials have expired and no Refresh Token was available to renew them."));
```

Comment:
Why not update the error message to denote that the credentials "need to be refreshed" instead of saying that they expired? That error will also trigger if the credentials have not expired, but the scope changed

Reply:
I'll use "Credentials need to be renewed but no Refresh Token is available to renew them."